tool/src/org/antlr/v4/codegen/Target.java

```diff
@@ -172,11 +177,20 @@ else if (shouldUseUnicodeEscapeForCodePointInDoubleQuotedString(c)) {
 		return buf.toString();
 	}
 
+	protected void appendUnicodeEscapedCodePoint(int codePoint, StringBuilder sb, boolean escape) {
+		if (escape) {
+			sb.append("\\");
+		}
+		appendUnicodeEscapedCodePoint(codePoint, sb);
+	}
+
 	/**
 	 * Escape the Unicode code point appropriately for this language
 	 * and append the escaped value to {@code sb}.
 	 */
-	abstract protected void appendUnicodeEscapedCodePoint(int codePoint, StringBuilder sb);
+	protected void appendUnicodeEscapedCodePoint(int codePoint, StringBuilder sb) {
```

Comment:
Ok, so now we group escaping to a central method but leave ability for target to override? makes sense but then we have a target overriding this and a diff implementation than what is in UnicodeEscapes.appendEscapedCodePoint. Isn't this what virtual method calls were all about? (removing IF/switches?)

Reply:
Seems like we should either do switch and call directly or use virtual methods but not both.

Reply:
The problem is in the static method `appendEscapedCodePoint`. It's used both in another static method `escapeCodePoint` and in `Target.appendUnicodeEscapedCodePoint`. Static `escapeCodePoint` is used in [tests](https://github.com/antlr/antlr4/blob/master/tool-testsuite/test/org/antlr/v4/test/tool/TestUnicodeEscapes.java#L19) and there is no access to Target in these tests.

```
@Test
public void latinJavaEscape() {
	checkUnicodeEscape("\\u0061", 0x0061, JavaTarget.key);
}
```

`Target` requires `CodeGenerator` that requires tool and grammar as constructor parameters. It's possible to use fake grammar, tool in tests for creating fake `CodeGenerator` and `Target` but it does not look clear.

That's why I've split methods in such a way.

Reply:
Good thoughts. yes. hmm...well, this argues for going completely to your switch but then we need an enum or const string.